cmd/clusterctl/phases/pivot.go

```diff
@@ -0,0 +1,284 @@
+	klog.Infof("Retrieving list of MachineDeployments to move for Cluster %s/%s", cluster.Namespace, cluster.Name)
+	machineDeployments, err := from.GetMachineDeploymentObjectsInNamespace(cluster.Namespace)
+	if err != nil {
+		return err
+	}
+	if err := moveMachineDeployments(from, to, machineDeployments); err != nil {
+		return err
+	}
+
+	klog.Infof("Retrieving list of MachineSets not associated with a MachineDeployment to move for Cluster %s/%s", cluster.Namespace, cluster.Name)
+	machineSets, err := from.GetMachineSetObjectsInNamespace(cluster.Namespace)
+	if err != nil {
+		return err
+	}
+	if err := moveMachineSets(from, to, machineSets); err != nil {
+		return err
+	}
+
+	klog.Infof("Retrieving list of Machines not associated with a MachineSet to move for Cluster %s/%s", cluster.Namespace, cluster.Name)
+	machines, err := from.GetMachineObjectsInNamespace(cluster.Namespace)
+	if err != nil {
+		return err
+	}
+	if err := moveMachines(from, to, machines); err != nil {
+		return err
+	}
+
+	if err := from.ForceDeleteCluster(cluster.Namespace, cluster.Name); err != nil {
+		return errors.Wrapf(err, "error force deleting cluster %s/%s", cluster.Namespace, cluster.Name)
+	}
+
+	klog.Infof("Successfully moved Cluster %s/%s", cluster.Namespace, cluster.Name)
+	return nil
+}
+
+func moveMachineDeployments(from, to clusterclient.Client, machineDeployments []*clusterv1.MachineDeployment) error {
+	for _, md := range machineDeployments {
+		if err := moveMachineDeployment(from, to, md); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func moveMachineDeployment(from, to clusterclient.Client, md *clusterv1.MachineDeployment) error {
+	klog.Infof("Moving MachineDeployment %s/%s", md.Namespace, md.Name)
+	klog.Infof("Retrieving list of MachineSets for MachineDeployment %s/%s", md.Namespace, md.Name)
+	machineSets, err := from.GetMachineSetsForMachineDeployment(md)
+	if err != nil {
+		return err
+	}
+	if err := moveMachineSets(from, to, machineSets); err != nil {
+		return err
+	}
+
+	// New objects cannot have a specified resource version. Clear it out.
+	md.SetResourceVersion("")
+	if err := to.CreateMachineDeploymentObjects([]*clusterv1.MachineDeployment{md}, md.Namespace); err != nil {
+		return errors.Wrapf(err, "error copying MachineDeployment %s/%s to target cluster", md.Namespace, md.Name)
+	}
+	if err := from.ForceDeleteMachineDeployment(md.Namespace, md.Name); err != nil {
+		return errors.Wrapf(err, "error force deleting MachineDeployment %s/%s from source cluster", md.Namespace, md.Name)
+	}
+	klog.Infof("Successfully moved MachineDeployment %s/%s", md.Namespace, md.Name)
+	return nil
+}
+
+func moveMachineSets(from, to clusterclient.Client, machineSets []*clusterv1.MachineSet) error {
+	for _, ms := range machineSets {
+		if err := moveMachineSet(from, to, ms); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func moveMachineSet(from, to clusterclient.Client, ms *clusterv1.MachineSet) error {
```

Comment:
@krousey I'm interested on your take here. Do you think it would be better to remove the controllerRefs and let the reconciliation process replace them, or would it make sense to go back and invert the logic again and just scale the target cluster controllers to 0 until the migration is complete?

Reply:
With further testing, the Machines are indeed correctly adopted by the MachineSets. I am currently trying to chase down an issue with MachineDeployments not properly handling cascading delete and orphaning their MachineSets after pivot, though.